### PR TITLE
Initial attempt at Docker package

### PR DESCRIPTION
**Pull request by michiels**
This PR contains an initial prototype to package Intercity as a Docker stack. A simple Dockerfile has been added for hosting regular Rails apps with asset compilation. A docker-compose.yml file is added so that you can set up a full Intercity + Redis + Postgres stack while developing, but this can also be used while deploying.

The deployment direction this PR introduces in the future would be as follows:

1. We publish intercity-next contaimer images to Docker Hub/store on every release.
2. We provide a docker-compose.yml file (possibily as part of intercity-cli) to launch an Intercity instance which will pull down Redis, Postgres and run Intercity + the Sidekiq worker inside Intercity container.

How you can already use this PR:

1. Install Docker for your platform. The official Docker Mac client is great.
2. Start a local development Stack via:

    ```
    docker-compose run web rake db:create
    docker-compose run web rake db:schema:load
    docker-compose up
    ```

3. Profit and view  your local Intercity instance on http://localhost:3000

Happy to further discuss how we could take this further and what additional tooling/scripts/CLIs we might want to offer for easy installation and upgrading. I can imagine that:

```
intercity-server install
```

can just use docker-compose under the hood to get an Intercity environment up&running on a server.

### CODE REVIEW

**Comment by jvanbaarsen** [comment]
@michiels Interesting! Thanks for taking a stab at this. How would this deal with the configuration that we currently do in the `app.yml` from the intercity-docker package? For example setting up Let's Encrypt etc?

**Comment by michiels** [comment]
@jvanbaarsen As far as I can see now, all the configuration that happens in `app.yml` (looking at https://github.com/intercity/intercity-docker/blob/master/samples/app.yml) can (or should) be done in `docker-compose.yml`, most importantly setting the ENV vars.

**Comment by michiels** [comment]
Ah! Maybe one discussion comment regarding nginx/letsencrypt setup. I think the Intercity Docker image in itself should stay clear of what's running in front of it. We can offer the the installation CLI to install nginx or configure nginx and/or letsencrypt. So not by configuring those inside the container itself, but by allowing the host to manage this itself.

That way an Intercity instance can be easily ran alongside Dokku for example.

**Comment by michiels** [comment]
I've added to the `docker-compose.yml` a setup that should be fully production-worthy, including persistent storage for PostgreSQL and Redis.

**Comment by jvanbaarsen** [comment]
@michiels One of the design criteria for me with intercity-docker (The current one) was that no dependencies outside of Docker should be installed on the server. For me the preference would remain to keep that in place.

File: config/sidekiq.yml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 ---
 :concurrency: 5
 :pidfile: tmp/sidekiq.pid
-:logfile: log/sidekiq.log
```
Comment (jvanbaarsen) [inline config/sidekiq.yml]: I think we want to log sidekiq as well right?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    image: postgres
+    volumes:
+      - pg-data:/var/lib/postgresql/data
+  redis:
+    image: redis
+    volumes:
+      - redis-data:/data
+  web:
+    build: .
+    command: bundle exec rails s
```
Comment (jvanbaarsen) [inline docker-compose.yml]: What web server does this use?

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      - REDIS_URL=redis://redis:6379
+      - SECRET_KEY_BASE=a
+      - RAILS_SERVE_STATIC_FILES=true
+      - HOSTNAME=intercity.example.com
+      - RAILS_LOG_TO_STDOUT=true
+    depends_on:
+      - db
+      - redis
+volumes:
+  pg-data:
```
Comment (jvanbaarsen) [inline docker-compose.yml]: Is this OK to be empty?

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    depends_on:
+      - db
+      - redis
+    environment:
+      - RAILS_ENV=production
+      - DATABASE_URL=postgresql://postgres:@db:5432/intercity
+      - SECRET_KEY_BASE=a
+      - RAILS_SERVE_STATIC_FILES=true
+      - REDIS_URL=redis://redis:6379
+      - HOSTNAME=intercity.example.com
```
Comment (jvanbaarsen) [inline docker-compose.yml]: Are these values overridable?

**Comment by michiels** [comment]
@jvanbaarsen Allright, we can discover that direction too by adding multiple dependencies to the same container. But you're going to have an nginx on the host system always if you want to allow multiple apps to be hosted on a server. So this new docker-compose thing takes the Docker philosophy into account that can also be launched on Docker Swarm setups. In the current `intercity-docker` it uses Docker more as a package manager instead of using Docker's fundamentals.

I would say we can just prototype various directions and see which one works best. We can also have two flavors run side-by-side.

File: docker-compose.yml (hunk 4 above)
Comment (michiels) [reply docker-compose.yml]: @jvanbaarsen Yep! You can specify them when executing the `docker-compose` command. Currently this file would only be suitable for development so some sane example defaults in there would be good.

File: docker-compose.yml (hunk 3 above)
Comment (michiels) [reply docker-compose.yml]: @jvanbaarsen Yeah, that indicates that it should just create a default Docker volume with no additional options.

File: docker-compose.yml (hunk 2 above)
Comment (michiels) [reply docker-compose.yml]: Puma by default.

**Comment by jvanbaarsen** [comment]
@michiels Yep sounds good. If it's ok with you I want to keep this PR open so I can test run this locally also. I probably have time for this next week.

File: config/sidekiq.yml (hunk 1 above)
Comment (michiels) [reply config/sidekiq.yml]: It does, but standard to STDOUT. This allows for proper logging to be picked up by other services in a standard containerized environment. We could allow configuring this via ENV vars.

Comment (jvanbaarsen) [reply config/sidekiq.yml]: 👍

File: docker-compose.yml (hunk 3 above)
Comment (jvanbaarsen) [reply docker-compose.yml]: :+1:

**Comment by michiels** [comment]
@jvanbaarsen Yes that's fine. This PR is not meant to be quickly reviewed and closed, but to propose a direction we can test over time.

File: docker-compose.yml (hunk 4 above)
Comment (jvanbaarsen) [reply docker-compose.yml]: 👍

**Comment by michiels** [comment]
@jvanbaarsen I think with this way, upgrading and installing Intercity is very easy. What I imagine a CLI to do:

1. Ask you (or show you how) to configure your setup by configuring the ENV vars
2. Run docker-compose to launch the Intercity stack
3. Set up (or use currently installed) an nginx with LetsEncrypt in front of it

We can still debate wether nginx should be run from inside the Intercity container or from a different container or on the host. That last option would be the ideal Dockerized way of working.

For the end user there would be no difference, other than perhaps the bonus to run Intercity alongside of other applications on the server.

**Comment by michiels** [comment]
@jvanbaarsen One thing that's quite cool to already try out is local development. I'll push a commit to a seperate branch in a bit where you can use this Docker Compose file for local development.

That means that someone only has to have Docker installed on their local machine, and have the Git repo checked oud. No need to install Redis, Postgres or even run bundler. You can just do `docker-compose run -P 3000:3000 web` and you will have an instant fully-fledged dev environment with Redis and PG setup.

**Pull request closed without merging**